File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
/*
    if (비어 있으면) push
    if (stack.peek()= '(' 이면) pop 하고 push
*/
import java.util.*;
class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (stack.isEmpty()) {
              stack.push(c);  
            } 
            
            else if (c == ')' && stack.peek() == '(') {
                stack.pop();
            }
            
            else {
                stack.push(c);
            }
        }
        
        return stack.isEmpty();
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    
    stack = []
    
    for i in s:
        if i == '(':
            stack.append(i)
        else:
            if not stack:
                return False
            else:
                stack.pop()
    
    if stack:
        return False

    return True